{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import io\n",
    "import sys\n",
    "import glob\n",
    "import enum\n",
    "import json\n",
    "import dask\n",
    "import xlrd\n",
    "import base64\n",
    "import imageio\n",
    "import requests\n",
    "import datetime\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import skimage\n",
    "\n",
    "import dask.diagnostics\n",
    "import sqlalchemy as db\n",
    "import sqlalchemy.orm\n",
    "import sqlalchemy.ext.declarative\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "\n",
    "sys.path.append('..')\n",
    "%aimport opencell.imaging.managers\n",
    "%aimport opencell.imaging.processors\n",
    "%aimport opencell.file_utils\n",
    "# %aimport opencell.database.operations\n",
    "\n",
    "from opencell import constants, file_utils\n",
    "from opencell.cli import database as db_cli\n",
    "from opencell.cli import imaging as imaging_cli\n",
    "from opencell.database import models\n",
    "from opencell.database import operations as ops\n",
    "from opencell.database import utils as db_utils\n",
    "from opencell.imaging import utils as im_utils\n",
    "from opencell.imaging import images, managers, processors, viz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'postgresql://postgres:password@cap.czbiohub.org:5434/opencelldb'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = db_utils.url_from_credentials('../db-credentials-dev.json')\n",
    "url = db_utils.url_from_credentials('../db-credentials-cap.json')\n",
    "url"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create and maybe populate the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually drop and create the schema\n",
    "engine = db.create_engine(url)\n",
    "\n",
    "im_sure = False\n",
    "if im_sure:\n",
    "    print('Dropping all tables')\n",
    "    models.Base.metadata.drop_all(engine)\n",
    "\n",
    "print('Creating all tables')\n",
    "models.Base.metadata.create_all(engine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop, create, and populate the database\n",
    "# TODO: fewer and less verbose warnings when drop_all=False\n",
    "im_sure = False\n",
    "if im_sure:\n",
    "    db_cli.populate(url, drop_all=False, errors='ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sanity checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "engine = db.create_engine(url)\n",
    "session_factory = db.orm.sessionmaker(bind=engine)\n",
    "Session = db.orm.scoped_session(session_factory)\n",
    "session = Session()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Session.query(models.CellLine).filter(models.CellLine.line_type=='PROGENITOR').first().name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of cell lines\n",
    "lines = session.query(models.CellLine).all()\n",
    "len(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the number of crispr designs\n",
    "designs = session.query(models.CrisprDesign).all()\n",
    "len(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all target_names\n",
    "names = [row.target_name for row in session.query(models.CrisprDesign).all()]\n",
    "len(names), len(set(names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[row.as_dict() for row in session.query(models.CrisprDesign).filter(models.CrisprDesign.target_name == 'c12orf66').all()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count the number of CrisprDesign rows per target_name\n",
    "d = pd.DataFrame(data=[names, np.ones((len(names),))]).transpose()\n",
    "d.columns = ['name', 'num']\n",
    "dn = d.groupby('name').count().reset_index().sort_values(by='num', ascending=False)\n",
    "dn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of facs and sequencing datasets\n",
    "facs = Session.query(models.FACSDataset).all()\n",
    "seq = Session.query(models.SequencingDataset).all()\n",
    "len(facs), len(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of fovs\n",
    "fovs = Session.query(models.MicroscopyFOV).all()\n",
    "len(fovs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = processors.FOVProcessor.from_database(fovs[-1])\n",
    "p.set_src_roots('/Users/keith.cheveralls/opencell-test/data/PlateMicroscopy/', '')\n",
    "p.src_filepath()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "profiles = ops.to_jsonable(p.calculate_z_profiles())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of fov results\n",
    "len(Session.query(models.MicroscopyFOVResult).all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# lines with FOVs\n",
    "lines = [line for line in Session.query(models.CellLine).all() if line.fovs]\n",
    "ops.PolyclonalLineOperations(lines[2]).get_top_scoring_fovs(session, ntop=1)[0].results[1].data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session.rollback()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# visualize the schema\n",
    "render_er(models.Base.metadata, '../test-schema.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download and cache metadata from UniprotKB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all target_names\n",
    "names = [row.target_name for row in session.query(models.CrisprDesign).all()]\n",
    "names = list(set(names))\n",
    "len(names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# url to retrieve the top hit for a given search string from human proteins in tab-delimited format\n",
    "# (note the explicit list of column names)\n",
    "url = (\n",
    "    'https://www.uniprot.org/uniprot/?'\n",
    "    'query=reviewed:yes+AND+organism:9606+AND+%s&sort=score&format=tab&limit=1&'\n",
    "    'columns=id,entry name,reviewed,protein names,genes,organism,length,comment(FUNCTION),families'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_uniprot(name):\n",
    "    response = requests.get(url % name)\n",
    "    if response.text:\n",
    "        return pd.read_csv(io.StringIO(response.text), sep='\\t')\n",
    "    else:\n",
    "        print('No result for %s' % name)\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tasks = [dask.delayed(get_uniprot)(name) for name in names]\n",
    "with dask.diagnostics.ProgressBar():\n",
    "    rows = dask.compute(*tasks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.concat(tuple(rows), axis=0)\n",
    "d.to_csv('/Users/keith.cheveralls/Downloads/2019-12-16_top-uniprotKB-hit-for-all-targets.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insert FACS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_cli.insert_facs(session, facs_results_dir='../../opencell-off-git/results/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insert microscopy datasets\n",
    "\n",
    "These are the pipeline-related ML IDs from the 'Microscopy Master Key' google sheet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_cli.insert_microscopy_datasets(session)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insert microscopy images\n",
    "\n",
    "Note that there are two MLs - ML0084 and ML0108 - that are not pipeline-related acquisitions but from which there are images in the PlateMicroscopy directory (and there fore in pm.md_raw)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cache_dir = '../plate-microscopy-cache/20191114-ess/'\n",
    "db_cli.insert_plate_microscopy_fovs(session, cache_dir=cache_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine how many FOVs were inserted (from the PlateMicroscopy directory only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pm.md_raw.shape, len(session.query(models.MicroscopyFOV).all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FOVs for controls are not inserted\n",
    "num_controls = pm.md_raw.loc[pm.md_raw.well_id.isin(['A01', 'H12'])].shape[0]\n",
    "num_controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FOVs from PML0084 and PML0108 are not inserted (because these acquistions were not truly pipeline)\n",
    "pml_ids = [row.pml_id for row in session.query(models.MicroscopyDataset).all()]\n",
    "num_nonpipeline = pm.md_raw.loc[~pm.md_raw.pml_id.isin(pml_ids)].shape[0]\n",
    "num_nonpipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the uninserted FOVs are likely the Jin samples that are not yet in the database\n",
    "# (these are mostly in Plate6 column E)\n",
    "pm.md_raw.shape[0] - num_controls - num_nonpipeline, len(session.query(models.MicroscopyFOV).all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fov = ops.PolyclonalLineOperations.from_plate_well(session, 'P0019', 'H11').cell_line.microscopy_fovs[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FOVs from raw-pipeline-microscopy datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = Session.query(models.MicroscopyDataset).filter(models.MicroscopyDataset.pml_id == 'PML0235').first()\n",
    "len(dataset.fovs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[r.kind for r in dataset.fovs[-1].results]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.fovs[0].results[1].data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insert FOV ROIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "src_root = '/Users/keith.cheveralls/opencell-test/data/PlateMicroscopy/'\n",
    "dst_root = '/Users/keith.cheveralls/opencell-test/output/opencell-microscopy/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21515, 'ATL2')"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fovs = session.query(models.MicroscopyFOV).all()\n",
    "p = processors.FOVProcessor.from_database(fovs[0])\n",
    "len(fovs), p.target_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p.crop_corner_rois(\n",
    "    '/Users/keith.cheveralls/opencell-test/data/PlateMicroscopy/', \n",
    "    '/Users/keith.cheveralls/opencell-test/output/opencell-microscopy/'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate FOV thumbnails"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "im_405 = imageio.imread(\n",
    "    '/Users/keith.cheveralls/image-data/raw-pipeline-microscopy/PML0227/QC/z-projections/MMStack_115-C2-7_C0-PROJ-Z.tif')\n",
    "im_488 = imageio.imread(\n",
    "    '/Users/keith.cheveralls/image-data/raw-pipeline-microscopy/PML0227/QC/z-projections/MMStack_115-C2-7_C1-PROJ-Z.tif')\n",
    "\n",
    "im_rgb = processors.FOVProcessor.make_rgb(im_405, im_488)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare downsampled images using rescale and downscale_local_mean\n",
    "viz.imshow(\n",
    "    np.concatenate(\n",
    "        (\n",
    "            skimage.transform.rescale(im_rgb, scale=.2, order=0, anti_aliasing=False, multichannel=True),\n",
    "            skimage.transform.rescale(im_rgb, scale=.2, order=1, anti_aliasing=False, multichannel=True),\n",
    "            skimage.transform.rescale(im_rgb, scale=.2, order=1, anti_aliasing=True, multichannel=True),\n",
    "            im_utils.autoscale(skimage.transform.downscale_local_mean(im_rgb, factors=(5, 5, 1)))[:, :, :].astype('float64')/255,\n",
    "        ), \n",
    "        axis=1),\n",
    "    figsize=14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fovs = session.query(models.MicroscopyFOV).all()\n",
    "p = processors.FOVProcessor.from_database(fovs[0])\n",
    "len(fovs), p.target_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = p.generate_fov_thumbnails('/Users/keith.cheveralls/opencell-test/output/opencell-microscopy/', scale=5, quality=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('test.json', 'w') as file:\n",
    "    json.dump(result, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with io.BytesIO(base64.b64decode(result['b64_strings']['rgb'])) as file:\n",
    "    im = imageio.imread(file)\n",
    "viz.imshow(im, figsize=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21507"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tns = Session.query(models.Thumbnail).filter(models.Thumbnail.channel == 'rgb').all()\n",
    "len(tns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[autoreload of opencell.imaging.processors failed: Traceback (most recent call last):\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 245, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 452, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 389, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 359, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 317, in update_instances\n",
      "    update_instances(old, new, obj, visited)\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 317, in update_instances\n",
      "    update_instances(old, new, obj, visited)\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 317, in update_instances\n",
      "    update_instances(old, new, obj, visited)\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 300, in update_instances\n",
      "    for obj in (obj for obj in objects if id(obj) not in visited):\n",
      "  File \"/anaconda3/envs/sklearnenv/lib/python3.7/site-packages/IPython/extensions/autoreload.py\", line 300, in <genexpr>\n",
      "    for obj in (obj for obj in objects if id(obj) not in visited):\n",
      "KeyboardInterrupt\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "d = [tns[ind].as_dict() for ind in range(1800)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('test.json', 'w') as file: \n",
    "    json.dump(d, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "146.65265084545862"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sqrt(21507)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21470"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "113*190"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_rows = 113\n",
    "n_cols = 190\n",
    "tile = np.zeros((n_rows*204, n_cols*204, 3), dtype='uint8')\n",
    "count = 0\n",
    "for ind in range(n_rows):\n",
    "    for jnd in range(n_cols):\n",
    "        with io.BytesIO(base64.b64decode(tns[count].data)) as file:\n",
    "            im = imageio.imread(file)\n",
    "        tile[ind*204:(ind + 1)*204, jnd*204:(jnd + 1)*204, :] = im\n",
    "        count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23052, 38760, 3)"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tile.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "imageio.imsave('tile_q90.jpg', tile[::2, ::2, :], quality=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viz.imshow(tile[::4, ::4,:], figsize=16)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inspect aggregated FOV results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Session.rollback()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_processing_events():\n",
    "    '''\n",
    "    This method is specific to aggregating processing events\n",
    "    because the JSON in the data column for processing events is a list, not a dict\n",
    "    '''\n",
    "    results = Session.query(models.MicroscopyFOVResult)\\\n",
    "        .filter(models.MicroscopyFOVResult.kind == 'raw-tiff-processing-events').all()  \n",
    "    data = [\n",
    "        [{\n",
    "            'fov_id': result.fov.id, \n",
    "            'line_id': result.fov.cell_line_id, \n",
    "            'pml_id': result.fov.dataset.pml_id,\n",
    "            **row\n",
    "        } for row in result.data] \n",
    "        for result in results\n",
    "    ]\n",
    "    rows = []\n",
    "    [rows.extend(row) for row in data]\n",
    "    df = pd.DataFrame(data=rows)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_results(kind):\n",
    "    '''\n",
    "    Aggregate results whose data column is a dict (not a list)\n",
    "    '''\n",
    "    results = Session.query(models.MicroscopyFOVResult)\\\n",
    "        .filter(models.MicroscopyFOVResult.kind == kind).all()  \n",
    "    data = [{\n",
    "        'fov_id': result.fov.id, \n",
    "        'line_id': result.fov.cell_line_id, \n",
    "        'pml_id': result.fov.dataset.pml_id,\n",
    "        **result.data\n",
    "    } for result in results]\n",
    "    df = pd.DataFrame(data=data)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_results_fast(kind):\n",
    "    query = '''\n",
    "        select fov.*, res.kind as kind, res.data as data from microscopy_fov fov\n",
    "        left join (select * from microscopy_fov_result where kind = '%s') res \n",
    "        on fov.id = res.fov_id;'''\n",
    "    df = pd.read_sql(query % kind, engine)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all processing events grouped by message\n",
    "df = all_processing_events()\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.groupby('message').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all z-profiles\n",
    "df = all_results('z-profiles')\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all cell-layer-alignment results\n",
    "df = all_results_fast('cell-layer-alignment')\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge the results JSON column into the dataframe\n",
    "df = df.merge(pd.DataFrame(data=list(df.data)), left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.isna(df.error).sum() / df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count the kinds of errors\n",
    "df.loc[~df.error.isna()].groupby('error').count().id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((18000, 13), 1755, (1723,))"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all FOV features\n",
    "df = all_results('fov-features')\n",
    "df.shape, len(set(df.line_id)), df.groupby('line_id').score.nlargest(1).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of cell lines with no score-able FOVs\n",
    "df.groupby('line_id').score.max().isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.score.isna().sum(), (df.score > -.5).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the four highest-scoring FOVs for each cell_line\n",
    "top4 = df.sort_values(by=['line_id', 'score'], ascending=False).groupby('line_id').head(4)\n",
    "top4.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.groupby('line_id').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insert sequencing results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SEQ_ROOT = '/Users/keith.cheveralls/Box/PipelineSequencing/CRISPRessoOUT_QC_Spreadsheets/'\n",
    "\n",
    "seq_sheet_filenames = {\n",
    "    1: \"mNGplate1REDO_sorted_CRISPResso_QC.xlsx\", \n",
    "    2: \"mNGplate2REDO_CRISPResso_QC.xlsx\", \n",
    "    3: \"mNGplate3REDO_HC_CRISPResso_QC.xlsx\", \n",
    "    4: \"mNGplate4REDO_sorted_CRISPResso_QC.xlsx\", \n",
    "    5: \"mNGplate5_CRISPResso_QC.xlsx\", \n",
    "    6: \"mNGplate6_CRISPResso_QC_HC.xlsx\", \n",
    "    7: \"mNGplate7_sorted_1to100_CRISPResso_QC.xlsx\",\n",
    "    8: \"mNGplate8_CRISPResso_QC.xlsx\",\n",
    "    9: \"mNGplate9_CRISPResso_QC.xlsx\",\n",
    "    10: \"mNGplate10_CRISPResso_QC.xlsx\",\n",
    "    11: \"mNGplate11_CRISPResso_QC.xlsx\",\n",
    "    12: \"mNGplate12_sorted_CRISPResso_QC.xlsx\",\n",
    "    13: \"mNGplate13sorted_CRISPResso_QC.xlsx\",\n",
    "    14: \"mNGplate14_CRISPResso_QC.xlsx\",\n",
    "    15: \"mNGplate15_CRISPResso_QC.xlsx\",\n",
    "    16: \"mNGplate16_CRISPResso_QC.xlsx\",\n",
    "    17: \"mNGplate17_CRISPResso_QC.xlsx\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_float(value):\n",
    "    try:\n",
    "        return float(value)\n",
    "    except ValueError:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: refactor this to use pandas to read the excel files\n",
    "def read_sequencing_sheet(filepath):\n",
    "    \n",
    "    sheet = xlrd.open_workbook(filepath).sheet_by_index(0)\n",
    "    num_rows = sheet.nrows\n",
    "    num_cols = sheet.ncols\n",
    "    \n",
    "    # HACK: hard-coded columns corresponding to final HDR ratios (HDR/all and HDR/modified)\n",
    "    WELL_ID_COLUMN_INDEX = 0\n",
    "    HDR_ALL_COLUMN_INDEX = num_cols - 2\n",
    "    HDR_MODIFIED_COLUMN_INDEX = num_cols - 1\n",
    "    \n",
    "    START_ROW_INDEX = 4\n",
    "    rows = range(START_ROW_INDEX, num_rows)\n",
    "    \n",
    "    # well_ids\n",
    "    well_ids = [sheet.cell_value(ind, WELL_ID_COLUMN_INDEX) for ind in rows]\n",
    "\n",
    "    # overall percent HDR\n",
    "    hdr_all = np.array([sheet.cell_value(ind, HDR_ALL_COLUMN_INDEX) for ind in rows])\n",
    "\n",
    "    # percent HDR of non-unmodified sequences\n",
    "    hdr_modified = np.array([sheet.cell_value(ind, HDR_MODIFIED_COLUMN_INDEX) for ind in rows])\n",
    "    \n",
    "    d = pd.DataFrame(\n",
    "        data=list(zip(well_ids, hdr_all, hdr_modified)),\n",
    "        columns=['well_id', 'hdr_all', 'hdr_modified'])\n",
    "    \n",
    "    # coerce to float\n",
    "    d['hdr_all'] = d.hdr_all.apply(to_float)\n",
    "    d['hdr_modified'] = d.hdr_modified.apply(to_float)\n",
    "    \n",
    "    # HACK: deal with missing/NaN values, \n",
    "    # which for some reason are loaded by xlrd as either the number 7 or 15\n",
    "    d['hdr_all'] = [val if val < 1 else None for val in d.hdr_all]\n",
    "    d['hdr_modified'] = [val if val < 1 else None for val in d.hdr_modified]\n",
    "        \n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load and concat all of the spreadsheets\n",
    "sequencing_data = []\n",
    "for plate_num, filename in seq_sheet_filenames.items():\n",
    "    d = read_sequencing_sheet(os.path.join(SEQ_ROOT, filename))\n",
    "    d['plate_num'] = plate_num\n",
    "    sequencing_data.append(d)\n",
    "sequencing_data = pd.concat(tuple(sequencing_data), axis=0)\n",
    "sequencing_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequencing_data.loc[sequencing_data.hdr_all.isna()].groupby('plate_num').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "engine.url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert into the database\n",
    "for ind, row in sequencing_data.iterrows():\n",
    "    if pd.isna(row.hdr_all):\n",
    "        continue\n",
    "        \n",
    "    plate_id = db_utils.format_plate_design_id(row.plate_num)\n",
    "    well_id = db_utils.format_well_id(row.well_id)\n",
    "    try:\n",
    "        pcl_ops = ops.PolyclonalLineOperations.from_plate_well(Session, plate_id, well_id)\n",
    "    except Exception as error:\n",
    "        print(str(error))\n",
    "        continue\n",
    "    \n",
    "    pcl_ops.insert_sequencing_dataset(Session, row[['hdr_all', 'hdr_modified']], errors='warn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(Session.query(models.SequencingDataset).all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sklearnenv",
   "language": "python",
   "name": "sklearnenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
